orbit/pkg/installer/installer.go

```diff
@@ -0,0 +1,251 @@
+	tempDirFn func(dir, pattern string) (string, error)
+
+	// execCmdFn can be set for tests to mock actual execution of the script. If
+	// nil, execCmd will be used, which has a different implementation on Windows
+	// and non-Windows platforms.
+	execCmdFn func(ctx context.Context, scriptPath string, env []string) ([]byte, int, error)
+
+	// can be set for tests to replace os.RemoveAll, which is called to remove
+	// the script's temporary directory after execution.
+	removeAllFn func(string) error
+
+	connectOsquery func(*Runner) error
+
+	scriptsEnabled func() bool
+}
+
+func NewRunner(client Client, socketPath string, scriptsEnabled func() bool) *Runner {
+	r := &Runner{
+		OrbitClient:       client,
+		osquerySocketPaht: socketPath,
+		scriptsEnabled:    scriptsEnabled,
+	}
+
+	return r
+}
+
+func (r *Runner) Run(config *fleet.OrbitConfig) error {
+	if r.connectOsquery == nil {
+		r.connectOsquery = connectOsquery
+	}
+
+	if err := r.connectOsquery(r); err != nil {
+		return fmt.Errorf("software installer runner connecting to osquery: %w", err)
+	}
+	return r.run(context.Background(), config)
+}
+
+func connectOsquery(r *Runner) error {
+	if r.OsqueryClient == nil {
+		osqueryClient, err := osquery.NewClient(r.osquerySocketPaht, 10*time.Second)
+		if err != nil {
+			log.Err(err).Msg("establishing osquery connection for software install runner")
+			return err
+		}
+
+		r.OsqueryClient = osqueryClient.Client
+	}
+
+	return nil
+}
+
+func (r *Runner) run(ctx context.Context, config *fleet.OrbitConfig) error {
+	var errs []error
+	for _, installerID := range config.Notifications.PendingSoftwareInstallerIDs {
+		if ctx.Err() == nil {
+			payload, err := r.installSoftware(ctx, installerID)
+			if err != nil {
+				errs = append(errs, err)
+			}
+			if err := r.OrbitClient.SaveInstallerResult(payload); err != nil {
+				errs = append(errs, fmt.Errorf("saving software install results: %w", err))
+			}
+		}
+	}
+	if len(errs) != 0 {
+		return errors.Join(errs...)
+	}
+
+	return nil
+}
+
+func (r *Runner) preConditionCheck(ctx context.Context, query string) (bool, string, error) {
+	res, err := r.OsqueryClient.Query(ctx, query)
+	if err != nil {
+		return false, "", fmt.Errorf("precondition check: %w", err)
+	}
+
+	if res.Status == nil {
+		return false, "", errors.New("no query status")
+	}
+
+	response, err := json.Marshal(res.Response)
+	if err != nil {
+		return false, "", fmt.Errorf("marshalling query response: %w", err)
+	}
+
+	if res.Status.Code != 0 {
+		// TODO(roberto): we can't return the error as the
+		// result because the back-end considers any non-empty
+		// string as a success.
+		// return false, fmt.Sprintf("osqueryd returned error (%d): %s", res.Status.Code, res.Status.Message), fmt.Errorf("non-zero query status: %d \"%s\"", res.Status.Code, res.Status.Message)
+		return false, "", fmt.Errorf("non-zero query status: %d \"%s\"", res.Status.Code, res.Status.Message)
+	}
+
+	if len(res.Response) == 0 {
```

Comment:
Not doing a full review yet, but seeing this, we json-marshal the osquery `Response` field on line 129 (which has this type: https://pkg.go.dev/github.com/osquery/osquery-go/gen/osquery#ExtensionPluginResponse), if it is nil or an empty slice, it would marshal to `null` or `[]` and wouldn't be empty for this check, is this ok? It looks like this could mean the query returned no result, but we'd treat it as having a result.

Reply:
Oh sorry, completely missed that this check was on `res.Response` and not on the json-marshaled value... :facepalm:

Reply:
The JSON marshalled value is just for the person reviewing the execution, so it shouldn't matter too much